### PR TITLE
Resolves building issues (WRF/WRF-chem) with Intel compilers (ifx/icx)

### PR DESCRIPTION
**Pull request by ahheo**
Resolving building issues with Intel compilers (ifx/icx) for WRF/WRF-chem

TYPE: bug fix

KEYWORDS: Missing/Wrong Prototypes

SOURCE: Changgui Lin

DESCRIPTION OF CHANGES:
Problem:
Problem: Most C code written long ago. No prototypes were used. See #1823. And, this pr is kind of an extension to #1823 addressing the same issue for building WRF-chem.

Solution:
Add missing prototypes; rearrange function order.

LIST OF MODIFIED FILES: list of changed files (use `git diff --name-status master` to get formatted list)
	modified:   arch/configure.defaults
	modified:   chem/KPP/kpp/kpp-2.1/src/code.c
	modified:   chem/KPP/kpp/kpp-2.1/src/code.h
	modified:   chem/KPP/kpp/kpp-2.1/src/code_c.c
	modified:   chem/KPP/kpp/kpp-2.1/src/code_f77.c
	modified:   chem/KPP/kpp/kpp-2.1/src/code_f90.c
	modified:   chem/KPP/kpp/kpp-2.1/src/code_matlab.c
	modified:   chem/KPP/kpp/kpp-2.1/src/gdata.h
	modified:   chem/KPP/kpp/kpp-2.1/src/gen.c
	modified:   chem/KPP/kpp/kpp-2.1/src/gen_org.c
	modified:   chem/KPP/kpp/kpp-2.1/src/kpp.c
	modified:   chem/KPP/kpp/kpp-2.1/src/scan.h
	modified:   chem/KPP/kpp/kpp-2.1/src/scan.l
	modified:   chem/KPP/kpp/kpp-2.1/src/scan.y
	modified:   chem/KPP/util/wkc/Makefile
	modified:   chem/KPP/util/wkc/change_chem_Makefile.c
	modified:   chem/KPP/util/wkc/compare_kpp_to_species.c
	modified:   chem/KPP/util/wkc/gen_kpp.c
	modified:   chem/KPP/util/wkc/gen_kpp_args_to_Update_Rconst.c
	modified:   chem/KPP/util/wkc/gen_kpp_interf_utils.c
	modified:   chem/KPP/util/wkc/gen_kpp_interface.c
	modified:   chem/KPP/util/wkc/gen_kpp_mech_dr.c
	modified:   chem/KPP/util/wkc/gen_kpp_utils.c
	modified:   chem/KPP/util/wkc/get_kpp_chem_specs.c
	modified:   chem/KPP/util/wkc/get_wrf_jvals.c
	modified:   chem/KPP/util/wkc/get_wrf_radicals.c
	modified:   chem/KPP/util/wkc/kpp_data.c
	modified:   chem/KPP/util/wkc/protos_kpp.h
	modified:   chem/KPP/util/wkc/registry_kpp.c
	modified:   chem/KPP/util/write_decomp/integr_edit.c
	modified:   compile
	modified:   external/RSL_LITE/buf_for_proc.c
	modified:   external/RSL_LITE/c_code.c
	modified:   external/RSL_LITE/cycle.c
	modified:   external/RSL_LITE/gen_comms.c
	modified:   external/RSL_LITE/period.c
	modified:   external/RSL_LITE/rsl_bcast.c
	modified:   external/RSL_LITE/rsl_lite.h
	modified:   external/RSL_LITE/rsl_malloc.c
	modified:   external/RSL_LITE/swap.c
	modified:   external/RSL_LITE/task_for_point.c
	modified:   external/io_grib1/MEL_grib1/FTP_getfile.c
	modified:   external/io_grib1/MEL_grib1/gribfuncs.h
	modified:   external/io_grib1/MEL_grib1/gribgetbds.c
	modified:   external/io_grib1/MEL_grib1/gribhdr2file.c
	modified:   external/io_grib1/MEL_grib1/init_enc_struct.c
	modified:   external/io_grib1/MEL_grib1/init_gribhdr.c
	modified:   external/io_grib1/MEL_grib1/init_struct.c
	modified:   external/io_grib1/MEL_grib1/ld_enc_input.c
	modified:   external/io_grib1/MEL_grib1/ld_enc_lookup.c
	modified:   external/io_grib1/MEL_grib1/ld_grib_origctrs.c
	modified:   external/io_grib1/MEL_grib1/map_lvl.c
	modified:   external/io_grib1/MEL_grib1/map_parm.c
	modified:   external/io_grib1/MEL_grib1/prt_badmsg.c
	modified:   external/io_grib1/MEL_grib1/prt_inp_struct.c
	modified:   external/io_grib1/MEL_grib1/set_bytes.c
	modified:   external/io_grib1/MEL_grib1/upd_child_errmsg.c
	modified:   external/io_grib1/grib1_routines.c
	modified:   external/io_grib1/grib1_util/read_grib.c
	modified:   external/io_grib1/grib1_util/read_grib.h
	modified:   external/io_grib1/grib1_util/write_grib.c
	modified:   external/io_grib1/gribmap.c
	modified:   external/io_grib1/gribmap.h
	modified:   external/io_grib1/trim.c
	new file:   external/io_grib1/trim.h
	modified:   external/io_grib_share/get_region_center.c
	modified:   external/io_grib_share/open_file.c
	modified:   frame/pack_utils.c
	modified:   tools/Makefile
	modified:   tools/gen_config.c
	modified:   tools/gen_interp.c
	modified:   tools/gen_irr_diag.c
	modified:   tools/gen_streams.c
	modified:   tools/gen_wrf_io.c
	modified:   tools/misc.c
	modified:   tools/protos.h
	modified:   tools/reg_parse.c
	modified:   tools/sym.c
	modified:   tools/sym.h
	modified:   tools/symtab_gen.c

### CODE REVIEW

File: arch/configure.defaults [hunk 1]
```diff
@@ -304,12 +304,12 @@ DESCRIPTION     =       INTEL ($SFC/$SCC)
...
-SCC             =       icc
-CCOMP           =       icc
-DM_FC           =       mpif90 -f90=$(SFC)
-DM_CC           =       mpicc -cc=$(SCC)
+OMPCC           =       # -qopenmp
+SFC             =       ifx # ifort
+SCC             =       icx # icc
+CCOMP           =       icx # icc
+DM_FC           =       mpiifx #mpif90 -f90=$(SFC)
+DM_CC           =       mpiicx #mpicc -cc=$(SCC)
```
Comment (weiwangncar) [inline arch/configure.defaults:312]: @ahheo Should not change this option. Add another option like it is in PR#[1893](https://github.com/wrf-model/WRF/pull/1893).

**Comment by weiwangncar** [comment]
@ahheo Can we assume that your changes are limited in chem/KPP/ directory, while the rest is identical to that in PR#1823?

File: arch/configure.defaults (hunk 1 above)
Comment (ahheo) [reply arch/configure.defaults:312]: OK. Well, what should I do with this PR for the next step? Should I convert it as a draft and make a new commit and submit this PR again?

Comment (weiwangncar) [reply arch/configure.defaults]: @ahheo Please continue to modify this PR.

Comment (ahheo) [reply arch/configure.defaults]: @weiwangncar I have made this modification following your suggestion: restoring the changed option and adding another option as PR #1893 , by another commit [7361882](https://github.com/wrf-model/WRF/pull/1942/commits/736188253339eacaf664a6d1e90f3cb8e679b115). Well, the commit still failed in WRF regression test, and I was not able to check the details. Sorry that this is my first time making a PR, and I am still learning the procedures to make things work.

Comment (weiwangncar) [reply arch/configure.defaults]: @ahheo The failure of the regression tests is not your fault. We currently have some trouble running the tests on the Cloud. We are working to resolve the issues.

**Comment by weiwangncar** [comment]
The regression test results:

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 23           24
    Number of Builds       : 60           57
    Number of Simulations  : 158           150        0
    Number of Comparisons  : 95           86        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None

**Comment by weiwangncar** [comment]
@ahheo It would be helpful if you identify the routines you modified in addition to those modified in PR#1823, and explain why you need to modify them.

**Comment by weiwangncar** [comment]
@ahheo I tried to use Intel oneAPI to compile WRF-Chem + KPP, it failed. It worked when compiling KPP only. Have you tried to compile WRF-Chem and KPP? I also verified that your modification still works for compiling WRF along.

**Comment by ahheo** [comment]
> @ahheo I tried to use Intel oneAPI to compile WRF-Chem + KPP, it failed. It worked when compiling KPP only. Have you tried to compile WRF-Chem and KPP? I also verified that your modification still works for compiling WRF along.

@weiwangncar, yes, I compiled with WRF_CHEM=1 together with WRF_KPP=1 using ifx / icx / mpiifx / mpiicx successfully, (and also successfully using ifx / icx / mpif90 -fc ifx / mpicc -cc icx). It would be nice if you can provide the compile log.

**Comment by weiwangncar** [comment]
@ahheo Here are the steps I took to compile your branch:
1. clone your branch and check out branch CL_test
2. on our HPC, use intel-oneapi/2023.0.0 module
3. set environtment variables:

```
setenv WRF_EM_CORE 1
setenv WRF_CHEM 1
setenv WRF_KPP 1
setenv FLEX_LIB_DIR /usr/lib64
setenv YACC '/usr/bin/yacc -d'
```
4. configure using option 78 (icx/ifx, dmpar)
5. compile using 'compile em_real' >& make78chemkpp.txt

Here is the log file. 
[make78chemkpp.txt](https://github.com/wrf-model/WRF/files/13594555/make78chemkpp.txt)

**Comment by ahheo** [comment]
> @ahheo Here are the steps I took to compile your branch:
> 
> 1. clone your branch and check out branch CL_test
> 2. on our HPC, use intel-oneapi/2023.0.0 module
> 3. set environtment variables:
> 
> ```
> setenv WRF_EM_CORE 1
> setenv WRF_CHEM 1
> setenv WRF_KPP 1
> setenv FLEX_LIB_DIR /usr/lib64
> setenv YACC '/usr/bin/yacc -d'
> ```
> 
>  4. configure using option 78 (icx/ifx, dmpar)
> 5. compile using 'compile em_real' >& make78chemkpp.txt
> 
> Here is the log file. [make78chemkpp.txt](https://github.com/wrf-model/WRF/files/13594555/make78chemkpp.txt)

@weiwangncar the compile problem is due to the failure in finding libfl.a. I found the following lines in your make log:
> No libfl.a in /usr/lib64                                                        
> check if FLEX_LIB_DIR environment variable is set correctly                     
> (FLEX_LIB_DIR should be the complete pathname of the FLEX library libfl.a)      
> OR: Enter full path to flex library on your system                              
> PROBLEM: libfl.a NOT FOUND IN
You should check if libfl.a located in /usr/lib64.

**Comment by weiwangncar** [comment]
@ahheo Thanks for pointing it out. True, /usr/lib64 does not have libfl.a, it has /usr/libfl.so. Let me find out more about this lib.

**Comment by weiwangncar** [comment]
@ahheo Replacing libfl.a by libfl.so allowed the code to compile.

**Comment by weiwangncar** [comment]
@ahheo You probably received an email as well. If you have not, the regression tests have failed. Mostly because it failed to compile for all.

**Comment by weiwangncar** [comment]
@ahheo The last change did not pass the regression tests. I have moved this to the develop branch to give both you and us more time to review and test the change.

**Comment by ahheo** [comment]
> @ahheo The last change did not pass the regression tests. I have moved this to the develop branch to give both you and us more time to review and test the change.

@weiwangncar Absolutely OK. It would be nice if you can tell more information why the regression tests failed.

**Comment by ahheo** [comment]
> @ahheo The last change did not pass the regression tests. I have moved this to the develop branch to give both you and us more time to review and test the change.

@weiwangncar  Absolutely OK. It would be nice if you can tell more information why the regression tests failed.

**Comment by weiwangncar** [comment]
@ahheo The problem is that no output was sent back. This can indicate serious problems. Have you succeeded in compiling the updated code on your local computer?

**Comment by ahheo** [comment]
> @ahheo The problem is that no output was sent back. This can indicate serious problems. Have you succeeded in compiling the updated code on your local computer?

@weiwangncar Yes, I successfully compiled the updated code on my own machine. So, something goes weird?

**Comment by jordanschnell** [comment]
@weiwangncar - seems no issues with old compile options. Have the regression test issues been resolved? I will make a small commit to restart the reg test. If pass, I can approve.

**Comment by weiwangncar** [comment]
@jordanschnell Thanks for starting a new tests. No, up to this point, we haven't addressed it. This PR should affect the new Intel oneAPI compilers (ifx/icx) only. Let's see if we get any test results back.

**Comment by weiwangncar** [comment]
The regression test results:

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 23           24
    Number of Builds       : 60           57
    Number of Simulations  : 158           150        0
    Number of Comparisons  : 95           86        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None

**Comment by jordanschnell** [comment]
@weiwangncar - I can compile with the old intel compilers, but do not have access to oneAPI

**Comment by weiwangncar** [comment]
@jordanschnell What about Chem+KPP with old Intel compiler?

**Comment by jordanschnell** [comment]
Yes, chem+KPP is successful with old intel compilers

**Comment by weiwangncar** [comment]
@islas @mgduda This PR looks ready to be merged. Any different opinion on this?

**Review by jordanschnell — APPROVED** [review]
yes, approved

File: tools/gen_irr_diag.c [hunk 2]
```diff
@@ -273,10 +273,10 @@ int irr_diag_scalar_indices( char *dirname )
      strcat( line,piece );
    }
    strcat( line," /)\n" );
-   fprintf( fp_inc,line );
+   fprintf( fp_inc,"%s",line );
    fprintf( fp_inc," \n");
 
-   for( i = 0; i < nChmOpts,rxt_cnt[i] > 0; i++ ) {
+   for( i = 0; i < nChmOpts & rxt_cnt[i] > 0; i++ ) {
```
Comment (islas) [inline tools/gen_irr_diag.c:279]: Is this supposed to be a boolean `&&`?

File: tools/gen_wrf_io.c [hunk 3]
```diff
@@ -85,7 +85,7 @@ gen_wrf_io2 ( FILE * fp , char * fname, char * structname , char * fourdname, no
   {
 
 
-    if ( p->ndims > 3 && ! p->node_kind & FOURD ) continue ; /* short circuit anything with more than 3 dims, (not counting 4d arrays) */
+    if ( p->ndims > 3 && (! p->node_kind) & FOURD ) continue ; /* short circuit anything with more than 3 dims, (not counting 4d arrays) */
```
Comment (islas) [inline tools/gen_wrf_io.c:88]:
```suggestion
    if ( p->ndims > 3 && ( (! p->node_kind) & FOURD ) ) continue ; /* short circuit anything with more than 3 dims, (not counting 4d arrays) */
```

**Comment by weiwangncar** [comment]
@ahheo Would you like to update the PR based on the suggestions made?

**Comment by benkirk** [comment]
Hi, I lead the user support group for Derecho, and I just wanted to say we will really value this contribution.  WRF-Chem has been quite painful, so far.

Can I also ask a naïve question, in hopes of getting pointed in a useful direction?
I see from the [KPP github site](https://github.com/KineticPreProcessor/KPP/releases) it is actively developed, with releases as recent as 2 months ago.  Yet as this PR correctly observes the version in WRF-Chem is 20+ years old.  I'm sure there is a good reason for that(?). Can anyone explain why??

Thanks again, this is great work.

**Comment by ahheo** [comment]
> @ahheo Would you like to update the PR based on the suggestions made?

@weiwangncar Sorry that for the late reply because I just come back from holidays to office today. The PR was updated as suggested. Please see and review the commit [911facc](https://github.com/wrf-model/WRF/pull/1942/commits/911facc5e3403046881343019a2f60e4c47809e3).

File: compile [hunk 4]
```diff
@@ -407,15 +407,30 @@ else
...
-  set comp = ( `grep "^SFC" configure.wrf | cut -d"=" -f2-` )
-  if      ( "$comp[1]" == "gfortran" ) then
-    gfortran --version
-  else if ( "$comp[1]" == "pgf90" ) then
-    pgf90 --version
-  else if ( "$comp[1]" == "ifort" ) then
-    ifort -V
+# remove comment text
+  set comp = ( `grep "^SFC" configure.wrf | cut -d"#" -f1 | cut -d"=" -f2-` )
+#  if      ( "$comp[1]" == "gfortran" ) then
```
Comment (islas) [inline compile:412]: The previous logic can be removed

File: tools/gen_wrf_io.c [hunk 5]
```diff
@@ -85,7 +85,7 @@ gen_wrf_io2 ( FILE * fp , char * fname, char * structname , char * fourdname, no
   {
 
 
-    if ( p->ndims > 3 && ! p->node_kind & FOURD ) continue ; /* short circuit anything with more than 3 dims, (not counting 4d arrays) */
+    if ( p->ndims > 3 && ( (! p->node_kind) ) & FOURD ) continue ; /* short circuit anything with more than 3 dims, (not counting 4d arrays) */
```
Comment (islas) [inline tools/gen_wrf_io.c:88]: I believe the first set of inner parenthesis should group `(! p->node_kind) & FOURD`

**Review by islas — CHANGES_REQUESTED** [review]

**Comment by weiwangncar** [comment]
@islas Can you take hopefully a final check on this PR?

**Review by islas — APPROVED** [review]

**Review by weiwangncar — APPROVED** [review]
For WRF-Chem since Jordan Schnell has approved this before.

**Comment by weiwangncar** [comment]
@ahheo If you could let us where you work, we can add it to the PR. Thanks for contributing to our model development effort.

**Comment by ahheo** [comment]
> If you could let us where you work, we can add it to the PR.
I work in National Space Science Center, CAS. Feel free to add it or not as I made the changes on my own :).

**Comment by weiwangncar** [comment]
@ahheo OK, thanks for letting us know.

File: tools/gen_interp.c [hunk 6]
```diff
@@ -377,7 +377,7 @@ fprintf(fp,"                  ngrid%%parent_grid_ratio, ngrid%%parent_grid_ratio
                strcpy( tmpstr , pp->interpu_aux_fields ) ;
 	     } else if ( down_path & FORCE_DOWN ) {
                /* by default, add the boundary and boundary tendency fields to the arg list */
-               if ( ! p->node_kind & FOURD ) {
+               if ( (! p->node_kind) & FOURD ) {
```
Comment (DWesl) [inline tools/gen_interp.c:380]:
```suggestion
               if ( ! (p->node_kind & FOURD) ) {
```
or
```suggestion
               if ( (~ p->node_kind) & FOURD ) {
```
`registry.h` has `#define FOURD 8` so this conditional will never be met (`!p->node_kind)` will be 0 or 1)

File: tools/gen_wrf_io.c [hunk 7]
```diff
@@ -85,7 +85,7 @@ gen_wrf_io2 ( FILE * fp , char * fname, char * structname , char * fourdname, no
   {
 
 
-    if ( p->ndims > 3 && ! p->node_kind & FOURD ) continue ; /* short circuit anything with more than 3 dims, (not counting 4d arrays) */
+    if ( p->ndims > 3 && ( (! p->node_kind) & FOURD ) ) continue ; /* short circuit anything with more than 3 dims, (not counting 4d arrays) */
```
Comment (DWesl) [inline tools/gen_wrf_io.c:88]:
```suggestion
    if ( p->ndims > 3 && ! (p->node_kind & FOURD) ) continue ; /* short circuit anything with more than 3 dims, (not counting 4d arrays) */
```
or
```suggestion
    if ( p->ndims > 3 && (~ p->node_kind) & FOURD ) continue ; /* short circuit anything with more than 3 dims, (not counting 4d arrays) */
```
`registry.h` has `#define FOURD 8` so this conditional will never be met (`!p->node_kind)` will be 0 or 1)

File: tools/reg_parse.c [hunk 8]
```diff
@@ -1058,7 +1058,7 @@ check_dimspecs()
 		  p->assoc_nl_var_s,p->name ) ;
 	  return(1) ;
         }
-        if ( ! q->node_kind & RCONFIG )
+        if ( (! q->node_kind) & RCONFIG )
```
Comment (DWesl) [inline tools/reg_parse.c:1061]:
```suggestion
        if ( ! (q->node_kind & RCONFIG) )
```
or
```suggestion
        if ( (~ q->node_kind) & RCONFIG )
```
I suspect `RCONFIG != 1`, so this conditional is unlikely to be met as-is.

File: tools/reg_parse.c [hunk 9]
```diff
@@ -1083,7 +1083,7 @@ check_dimspecs()
 		p->assoc_nl_var_e,p->name ) ;
 	return(1) ;
       }
-      if ( ! q->node_kind & RCONFIG )
+      if ( (! q->node_kind) & RCONFIG )
```
Comment (DWesl) [inline tools/reg_parse.c:1086]:
```suggestion
      if ( ! (q->node_kind & RCONFIG) )
```
or
```suggestion
      if ( (~ q->node_kind) & RCONFIG )
```
I suspect `RCONFIG != 1`, so this conditional will never be met as-is.

**Review by DWesl — COMMENTED** [review]
While I'm glad you're clarifying the precedence of the bit-mask and logical-not operators, I have some questions about the choice of how to do so, and would appreciate hearing the reasons for responding to the warning by insisting you meant the behavior the compiler thought unlikely?

**Comment by ahheo** [comment]
> While I'm glad you're clarifying the precedence of the bit-mask and logical-not operators, I have some questions about the choice of how to do so, and would appreciate hearing the reasons for responding to the warning by insisting you meant the behavior the compiler thought unlikely?

@DWesl Thank you for raising the questions. I would point out that my corresponding changes focused only on the grouping of if-conditions but not the precedence of the bit-mask and logical-not operators. Assuming that the original code yields warnings but gives the right results, the rule applied for grouping is to have the same result after grouping. Indeed, I have made the following test (currently with only icx compiler):

> a is 1 and b is 0:
>  ! a & b is False
>  ! (a & b) is True
>  (! a) & b is False
> 
> a is 0 and b is 0:
>  ! a & b is False
>  ! (a & b) is True
>  (! a) & b is False
> 
> a is 0 and b is 1:
>  ! a & b is True
>  ! (a & b) is True
>  (! a) & b is True
> 
> a is 1 and b is 1:
>  ! a & b is False
>  ! (a & b) is False
>  (! a) & b is False

But I cannot confirm that the assumption is reasonable, and I wouldn't say if the warnings are related to the ifx compiler. Finally, it would be very appreciated if you can validate the assumption and/or made changes with your reasons provided.

**Comment by DWesl** [comment]
`FOURD = 8`, and `RCONFIG = 4`:
https://github.com/wrf-model/WRF/blob/794843fe72a92b373401660625bdcc08e1fd8409/tools/registry.h#L27-L28

I think if you use either value for `b`, the first and third results will always be false, even if you use 8, 12, or even 255 for `a`, at which point we may as well delete the conditional and make any `else` branch evaluate every time.  Phrased another way, I think the compiler warning is correct that whoever wrote the code was wrong about the relative precedence of `!` and `&`.

**Comment by ahheo** [comment]
> I think if you use either value for `b`, the first and third results will always be false, even if you use 8, 12, or even 255 for `a`
> I think the compiler warning is correct that whoever wrote the code was wrong about the relative precedence of `!` and `&`.
@DWesl I agree with you. Well, I have not idea what I need to do at this point as the pull request has been merged already.

**Pull request merged**